ablateLibrary/flow/compressibleFlow.cpp

```diff
@@ -1,34 +1,260 @@
+
+            // Get the speed of sound from the eos
+            PetscReal ie;
+            PetscReal a;
+            PetscReal p;
+            flowParameters->decodeStateFunction(NULL, dim, rho, xc[RHOE] / rho, vel, &ie, &a, &p, flowParameters->decodeStateFunctionContext) >> checkError;
+
+            PetscReal u = xc[RHOU] / rho;
+            PetscReal dt = flowParameters->cfl * dx / (a + PetscAbsReal(u));
+            dtMin = PetscMin(dtMin, dt);
+        }
+    }
+    PetscInt rank;
+    MPI_Comm_rank(PetscObjectComm((PetscObject)ts), &rank);
+
+    PetscReal dtMinGlobal;
+    MPI_Allreduce(&dtMin, &dtMinGlobal, 1, MPIU_REAL, MPI_MIN, PetscObjectComm((PetscObject)ts)) >> checkMpiError;
+
+    TSSetTimeStep(ts, dtMinGlobal) >> checkError;
+
+    if (PetscIsNanReal(dtMinGlobal)) {
+        throw std::runtime_error("Invalid timestep selected for flow");
+    }
+
+    VecRestoreArrayRead(v, &x) >> checkError;
 }
 
-void ablate::flow::CompressibleFlow::SetupSolve(TS &ts) {
-    // finish setup and assign flow field
-    CompressibleFlow_CompleteProblemSetup(flowData, ts);
-    ablate::flow::Flow::SetupSolve(ts);
+PetscErrorCode ablate::flow::CompressibleFlow::CompressibleFlowRHSFunctionLocal(DM dm, PetscReal time, Vec locXVec, Vec globFVec, void* ctx) {
+    PetscFunctionBeginUser;
+    PetscErrorCode ierr;
+
+    ablate::flow::CompressibleFlow* flow = (ablate::flow::CompressibleFlow*)ctx;
+
+    // compute the euler flux across each face (note CompressibleFlowComputeEulerFlux has already been registered)
+    ierr = DMPlexTSComputeRHSFunctionFVM(dm, time, locXVec, globFVec, &flow->compressibleFlowData);
+    CHKERRQ(ierr);
+
+    // if there are any coefficients for diffusion, compute diffusion
+    if (flow->compressibleFlowData->k || flow->compressibleFlowData->mu) {
+        // update any aux fields
+        ierr = FVFlowUpdateAuxFieldsFV(flow->dm->GetDomain(), flow->auxDM, time, locXVec, flow->auxField, TOTAL_COMPRESSIBLE_AUX_COMPONENTS, flow->auxFieldUpdateFunctions, flow->compressibleFlowData);
+        CHKERRQ(ierr);
+
+        // compute the RHS sources
+        ierr = CompressibleFlowDiffusionSourceRHSFunctionLocal(dm, flow->auxDM, time, locXVec, flow->auxField, globFVec, flow->compressibleFlowData);
+        CHKERRQ(ierr);
+    }
+
+    PetscFunctionReturn(0);
+}
+
+void ablate::flow::CompressibleFlow::CompleteProblemSetup(TS ts) {
+    Flow::CompleteProblemSetup(ts);
+
+    if (compressibleFlowData->automaticTimeStepCalculator) {
+        preStepFunctions.push_back(ComputeTimeStep);
+    }
+
+    // Override the DMTSSetRHSFunctionLocal in DMPlexTSComputeRHSFunctionFVM with a function that includes euler and diffusion source terms
+    DMTSSetRHSFunctionLocal(dm->GetDomain(), CompressibleFlowRHSFunctionLocal, this) >> checkError;
+
+    // copy over any boundary information from the dm, to the aux dm and set the sideset
+    if (auxDM) {
+        PetscDS flowProblem;
+        DMGetDS(dm->GetDomain(), &flowProblem) >> checkError;
```

Comment:
Note that technically a DM can have more than one DS. However, if ABLATE is not going to use that, then we can ignore it.